File: 2020/day10/main.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <fstream>
#include <numeric>
#include <vector>

using std::adjacent_difference;
using std::cout;
using std::endl;
using std::istream_iterator;
using std::ifstream;
using std::sort;
using std::vector;

int main(void) {
    ifstream f("input");
    assert(f.is_open());
    vector<unsigned long> values(istream_iterator<unsigned long>(f), istream_iterator<unsigned long>{});

    // puzzle 1
    sort(values.begin(), values.end());
    values.insert(values.begin(), 0);
    unsigned long max = *(values.end() - 1) + 3;
    values.push_back(max);
    vector<unsigned long> diffs(values.size());
    adjacent_difference(values.begin(), values.end(), diffs.begin());
    size_t ones = count(diffs.begin(), diffs.end(), 1);
    size_t threes = count(diffs.begin(), diffs.end(), 3);
    cout << ones * threes << endl;

    /// puzzle 2
    vector<unsigned long> S(max + 1);
    S[0] = 1;
    for (const auto& value: values) {
        for (unsigned long diff = 1; diff < 4; diff++) {
            if (diff <= value) {
                S[value] += S[value - diff];
            }
        }
    }
    cout << *(S.end() - 1) << endl;
}

File: 2020/day10/main.py
def main() -> None:
    with open('input') as f:
        values = [int(line) for line in f]

    # puzzle 1
    values.sort()
    values.insert(0, 0)
    values.append(values[-1] + 3)
    diffs = [values[i + 1] - values[i] for i in range(len(values) - 1)]
    print(diffs.count(1) * diffs.count(3))

    # puzzle 2
    S = [0] * (values[-1] + 1)
    S[0] = 1
    for value in values:
        for diff in [1, 2, 3]:
            if diff <= value:
                S[value] += S[value - diff]
    print(S[-1])


if __name__ == '__main__':
    main()
